lib/proc-macros/src/content_hash.rs

```diff
@@ -45,6 +48,77 @@ pub fn generate_hash_impl(data: &Data) -> TokenStream {
                 quote! {}
             }
         },
-        _ => unimplemented!("ContentHash can only be derived for structs."),
+        // Generates a match statement with a match arm and hash implementation
+        // for each of the variants in the enum.
+        Data::Enum(ref data) => {
+            let match_hash_statements =
+                data.variants
+                    .iter()
+                    .enumerate()
+                    .map(|(i, v)| match &v.fields {
+                        Fields::Named(fields) => named_fields_match_arm(i, v, fields),
+                        Fields::Unnamed(fields) => anonymous_fields_match_arm(i, v, fields),
+                        Fields::Unit => {
+                            let ix = index_to_ordinal(i);
+                            quote_spanned! {v.span() =>
+                                Self::#v => {state.update(&#ix.to_le_bytes());}
+                            }
+                        }
+                    });
+            quote! {
+                match self {
+                    #(#match_hash_statements)*
+                }
+            }
+        }
+        Data::Union(_) => unimplemented!("ContentHash cannot be derived for unions."),
+    }
+}
+
+// The documentation for `ContentHash` specifies that the hash impl for each
+// enum variant should hash the ordinal number of the enum variant as a little
+// endian u32 before hashing the variant's fields, if any.
+fn index_to_ordinal(ix: usize) -> u32 {
+    u32::try_from(ix).expect("The number of enum variants overflows a u32.")
+}
+
+fn enum_bindings<'a>(fields: impl IntoIterator<Item = &'a Field>) -> Vec<Ident> {
```

Comment:
nit: this really feels like two different functions without much in common and a single callsite each. Maybe just inline?

Reply:
It started out as a single function, but then the brace arms had to be different and I split it up. I agree we can just inline at this point since the bulk of the shared logic is in the `enum_bindings` function.